src/commands/play.rs

```diff
@@ -145,37 +171,68 @@ pub async fn play(
 
                 update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
             }
-            QueryType::PlaylistLink => {
-                if let Some(urls) = YouTubeRestartable::ytdl_playlist(url, mode).await {
-                    let mut insert_idx = 1;
-                    let mut is_first_playlist_track = true;
-
-                    for url in urls.into_iter() {
-                        insert_track(&call, url, QueryType::VideoLink, insert_idx)
-                            .await
-                            .unwrap();
-
-                        if is_first_playlist_track && !queue_was_empty {
-                            force_skip_top_track(&call.lock().await).await;
-                            is_first_playlist_track = false;
-                        } else {
-                            insert_idx += 1;
-                        }
-
-                        update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
+            QueryType::PlaylistLink(url) => {
+                let urls = YouTubeRestartable::ytdl_playlist(&url, mode)
+                    .await
+                    .ok_or(ParrotError::Other("failed to fetch playlist"))?;
+
+                let mut insert_idx = 1;
+                let mut is_first_playlist_track = true;
+
+                for url in urls.into_iter() {
+                    insert_track(&call, &QueryType::VideoLink(url), insert_idx)
+                        .await
+                        .unwrap();
+
+                    if is_first_playlist_track && !queue_was_empty {
+                        force_skip_top_track(&call.lock().await).await;
+                        is_first_playlist_track = false;
+                    } else {
+                        insert_idx += 1;
+                    }
```

Comment:
Hmm. Nah.

I would say before the for loop manually enqueue the first track and force the skip and then process the rest (you can use vector slicing almost like in python)

Reply:
Eh. then we'd have the same logic inside and outside the loop: enqueuing, skipping and updating the queue message.
I can ditch the flag with enumerate, though?

```rust
let mut insert_idx = 1;

for (i, url) in urls.into_iter().enumerate() {
    insert_track(&call, &QueryType::VideoLink(url), insert_idx)
        .await
        .unwrap();

    if i == 0 && !queue_was_empty {
        force_skip_top_track(&call.lock().await).await;
    } else {
        insert_idx += 1;
    }

    update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
}
```

Reply:
Coded this on GitHub markdown, probably does not compile but tweak it:

```rust
insert_track(&call, &QueryType::VideoLink(urls[0]), 1).await.unwrap();
if !queue_was_empty {
  force_skip_top_track(&call.lock().await).await;
}
update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;

for (i, url) in urls[1..].into_iter().enumerate() {
    insert_track(&call, &QueryType::VideoLink(url), i + 2)
        .await
        .unwrap();
    update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
}
```

Reply:
```rust
// insert first track, skipping to it if nothing was playing
insert_track(&call, &QueryType::VideoLink(url), 1).await.unwrap();
if !queue_was_empty {
    force_skip_top_track(&call.lock().await).await
}
update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await

// insert remaining tracks
for (i, url) in urls.into_iter().enumerate() {
    insert_track(&call, &QueryType::VideoLink(url), i + 2 - (queue_was_empty as usize)).await.unwrap();
    update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
}
```

Reply:
@joao-conde Do you think this is clearer? I'm torn.

Reply:
No, keep yours.